torch/lib/c10d/reducer.cpp

```diff
@@ -455,31 +520,38 @@ void Reducer::autograd_hook(VariableIndex index) {
     local_used_maps_[index.replica_index][index.variable_index] = 1;
   }
 
-  // Rebuild bucket only if 1) it is the first time to rebuild bucket 2)
-  // find_unused_parameters_ is false, currently it does not support when there
-  // are unused parameters 3) this backward pass needs to run allreduce. Here,
-  // we just dump tensors and their parameter indices into rebuilt_params_ and
-  // rebuilt_param_indices_ based on gradient arriving order, and then at the
-  // end of finalize_backward(), buckets will be rebuilt based on
-  // rebuilt_params_ and rebuilt_param_indices_, and then will be broadcasted
-  // and initialized. Also we only need to dump tensors and parameter indices of
-  // one replica.
-  push_rebuilt_params(index);
+  if (static_graph_ && num_iterations_ == 1) {
+    numGradHooksTriggeredMap_[index] += 1;
+    return;
```

Comment:
IIUC this is what prevents buckets from being marked as ready and then go to allreduce in the first iteration for static graph training. Hopefully it's not too prone to breaking via a refactor or anything like that. Maybe it's worth it to add `TORCH_CHECK(!static_graph_training_ || num_iterations > 1)` below to ensure we don't get to the actual mark variable ready logic when we should be running delay allreduce?